vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java

```diff
@@ -0,0 +1,1060 @@
+        List<Component> flatOrderedComponents = new ArrayList<>();
+        for (Component component : childrenComponents) {
+            flatOrderedComponents.add(component);
+            String itemRepresentation;
+            if (component instanceof DashboardSection section) {
+                flatOrderedComponents.addAll(section.getWidgets());
+                List<DashboardWidget> sectionWidgets = section.getWidgets();
+                int sectionIndex = itemIndex.getAndIncrement();
+                String sectionWidgetsRepresentation = sectionWidgets.stream()
+                        .map(widget -> getWidgetRepresentation(widget,
+                                itemIndex.getAndIncrement()))
+                        .collect(Collectors.joining(","));
+                itemRepresentation = "{ component: $%d, items: [ %s ], id: %d }"
+                        .formatted(sectionIndex, sectionWidgetsRepresentation,
+                                section.getElement().getNode().getId());
+            } else {
+                itemRepresentation = getWidgetRepresentation(
+                        (DashboardWidget) component,
+                        itemIndex.getAndIncrement());
+            }
+            itemRepresentations.add(itemRepresentation);
+        }
+        String updateItemsSnippet = "this.items = [ %s ];"
+                .formatted(String.join(",", itemRepresentations));
+        getElement().executeJs(updateItemsSnippet,
+                flatOrderedComponents.toArray(Component[]::new));
+    }
+
+    private void setI18nWithJS() {
+        JsonObject i18nJson = (JsonObject) JsonSerializer.toJson(i18n);
+
+        // Remove properties with null values to prevent errors in web
+        // component
+        removeNullValuesFromJsonObject(i18nJson);
+
+        // Assign new I18N object to WC, by merging the existing
+        // WC I18N, and the values from the new DashboardI18n instance,
+        // into an empty object
+        getElement().executeJs("this.i18n = Object.assign({}, this.i18n, $0);",
+                i18nJson);
+    }
+
+    private void removeNullValuesFromJsonObject(JsonObject jsonObject) {
+        for (String key : jsonObject.keys()) {
+            if (jsonObject.get(key).getType() == JsonType.NULL) {
+                jsonObject.remove(key);
+            }
+        }
+    }
+
+    private static String getWidgetRepresentation(DashboardWidget widget,
+            int itemIndex) {
+        return "{ component: $%d, colspan: %d, rowspan: %d, id: %d  }"
+                .formatted(itemIndex, widget.getColspan(), widget.getRowspan(),
+                        widget.getElement().getNode().getId());
+    }
+
+    private void doRemoveAll() {
+        new ArrayList<>(childrenComponents).forEach(child -> {
+            if (child instanceof DashboardSection section) {
+                doRemoveSection(section);
+            } else {
+                doRemoveWidget((DashboardWidget) child);
+            }
```

Comment:
Looks like there is no functional difference between removing sections or items, so a single method that can handle both seems enough.

Reply:
Combined both under the name `doRemoveItem`.